x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -853,13 +896,43 @@ private static LogicalPlan maybePushDownPastUnary(Filter filter, UnaryPlan unary
      *
      * Ordering the evals before the orderBys has the advantage that it's always possible to order the plans like this.
      * E.g., in the example above it would not be possible to put the eval after the two orderBys.
+     *
+     * In case one of the eval's fields would shadow the orderBy's attributes, we rename the attribute first.
+     *
+     * E.g.
+     *
+     * ... | sort a | eval a = b + 1 | ...
+     *
+     * becomes
+     *
+     * ... | eval $$a = a | eval a = b + 1 | sort $$a | drop $$a
      */
     protected static class PushDownEval extends OptimizerRules.OptimizerRule<Eval> {
         @Override
         protected LogicalPlan rule(Eval eval) {
             LogicalPlan child = eval.child();
 
             if (child instanceof OrderBy orderBy) {
```

Comment:
This method looks to be the same across all rules - better to extract it in a separate method and look to incorporate the renameAttributesInPlace in it as well if it makes sense.